### DISCUSSION TITLE
Advice for real-time rendering up to 10M points (zooming and panning)

### DISCUSSION THREAD

**Discussion by NataliaDSmirnova** [visgl/deck.gl#7172] [Q&A] [answered]
Hi, I would like to ask for some advice on best performance practice for rendering more than 1M points (up to 10M) in real-time using deck.gl. My experience at the moment gives two approaches:
1.  Adaptive data representation with two layers for different zoom levels:
    * Layer with GL_POINTS when zoom out
    * ScatterplotLayer when zoom in
2. Adaptive rendering. It seems like image improvement (detailing) after transformation stops.
    * Rendering only a data subset while panning or zooming
    * Rendering all data after transformation ends

Maybe there are other decisions? I appreciate any ideas :)

**Comment by ibgreen** [comment, marked as answer]
1) While it can vary on different GPUs and drivers, our past testing rendering with GL.POINTS did not provide performance benefits. It may be that you are fragment shader bound and that reducing the radius of points helps. 
2) While you can do numerous hacks such as the ones you suggest to push the limit a little bit, the ultimate solution that scales to any size point cloud is to tile your cloud (e.g. into 3D tiles) and render it using the 3D tile layer. 

For 3D tiling, you'd need to research tiling tools. Cesium has the ION service which is excellent if that is an option. There was an effort to build an open source point cloud tiler in loaders.gl but the maintainer moved on and I am not sure if it still works: https://github.com/visgl/loaders.gl/tree/master/cpp/Model3DTiler

**Reply by ibgreen** [reply]
Sorry almost forgot: COPC is the latest solution to this problem, and it looks very promising, though we haven't added support to deck.gl / loaders.gl yet: https://github.com/copcio/copcio.github.io

**Reply by NataliaDSmirnova** [reply]
Thank you so much! We are going to investigate references 👍

**Reply by Hutch07** [reply]
There is a repo from https://github.com/connormanning/copc.js .  Is there a way to package and pass the data to deck.gl ?

**Reply by wiesehahn** [reply]
has there been any progress related to COPCs?

**Reply by ibgreen** [reply]
So far no updates, unfortunately.
Some work on a COPCSource by a contributor was landed I think, but not integrated in a way that would be usable.

**Reply by do-me** [reply]
Just fyi that with WebGPU Scatterplot Layer I can visualize up to 260M points on a 128Gb M3 Max in Chrome.
GitHub: https://github.com/do-me/deck.gl-scatterplot-webgpu
Demo: https://do-me.github.io/deck.gl-scatterplot-webgpu/
WebGPU support in deck.gl is still in alpha but still works great!
